### DISCUSSION TITLE
Add support for CAN bus for ESP32

### DISCUSSION THREAD

**Discussion by robota10** [nanoframework/Home#1346] [General]
**Description**

2>Getting things ready to deploy assemblies to .NET nanoFramework device: ESP32_BLE_REV3 @ COM13.
2>Deploy failed.
2>
2>***************************************************************************
2>
2>
2>The connected target does not have support for the following assembly(ies):
2>
2>    'nanoFramework.Device.Can'
2>
2>Please check: 
2>  1) if the target is running the most updated image.
2>  2) if the target image was built to include support for all referenced assemblies.
2>
2>Our Visual Studio FAQ has a troubleshooting guide: https://docs.nanoframework.net/content/faq/working-with-vs-extension.html
2>
2>***************************************************************************
2>

System Information
HAL build info: nanoCLR running @ ESP32 built with ESP-IDF ac5d805
  Target:   ESP32_BLE_REV3
  Platform: ESP32

Firmware build Info:
  Date:        Aug  8 2023
  Type:        MinSizeRel build, chip rev. 3, support for PSRAM, support for BLE
  CLR Version: 1.8.1.423
  Compiler:    GNU ARM GCC v8.4.0

**How to solve the problem**

Add native implementation of Devices.Can.

**Comment by alberk8** [comment]
Have you tried this [nuget](https://www.nuget.org/packages/nanoFramework.Device.Can)?

**Reply by robota10** [reply]
yes.

![image](https://github.com/nanoframework/Home/assets/142675005/afc64694-70c2-4915-a238-e036de2228a0)

**Reply by robota10** [reply]
I found out from [here](https://github.com/nanoframework/nf-interpreter/pull/1087) that support for CAN bus was added about 5 years ago, but it only supports STM32, not ESP32. There seems to be no more news about it since.

**Reply by alberk8** [reply]
You cannot install the CAN bus wires directly into esp32, it will short the esp32 as the can bus is 12v (esp32 max input is 3.3v) , nF provide the protocol decoding not the transport medium, For that you still need the CAN transceiver. 

BTW, I check the code again.  the native side is not implemented.

**Reply by Ellerbach** [reply]
Correct, native side is not implemented. But it can be done as indeed, IDF supports it. And indeed, levelers will be needed not to grill the ESP32! I've commented with the link on the IDF functions in another issue
